sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnSignature.java

```diff
@@ -0,0 +1,275 @@
+import static org.apache.beam.sdk.transforms.DoFn.StartBundle;
+import static org.apache.beam.sdk.transforms.DoFnReflector.format;
+import static org.apache.beam.sdk.transforms.DoFnReflector.formatType;
+
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+
+import com.google.common.reflect.TypeParameter;
+import com.google.common.reflect.TypeToken;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Describes the signature of a {@link DoFn}, in particular, which features it uses, which extra
+ * context it requires, types of the input and output elements, etc.
+ *
+ * <p>See <a href="https://s.apache.org/a-new-dofn">A new DoFn</a>.
+ */
+class DoFnSignature {
+  public final TypeToken<?> inputT;
+  public final TypeToken<?> outputT;
+  public final ProcessElementMethod processElement;
+  public final BundleMethod startBundle;
+  public final BundleMethod finishBundle;
+
+  private DoFnSignature(
+      TypeToken<?> inputT,
+      TypeToken<?> outputT,
+      ProcessElementMethod processElement,
+      BundleMethod startBundle,
+      BundleMethod finishBundle) {
+    this.inputT = inputT;
+    this.outputT = outputT;
+    this.processElement = processElement;
+    this.startBundle = startBundle;
+    this.finishBundle = finishBundle;
+  }
+
+  /** Describes a {@link DoFn.ProcessElement} method. */
+  public static class ProcessElementMethod {
+    /** The relevant method in the user's class. */
+    public final Method targetMethod;
+    /** 0-based index of the {@link BoundedWindow} parameter, or -1 if it's not used. */
+    public final int boundedWindowParamIndex;
+    /** 0-based index of the {@link DoFn.InputProvider} parameter, or -1 if it's not used. */
+    public final int inputProviderParamIndex;
+    /** 0-based index of the {@link DoFn.OutputReceiver} parameter, or -1 if it's not used. */
+    public final int outputReceiverParamIndex;
+
+    private ProcessElementMethod(
+        Method targetMethod,
+        int boundedWindowParamIndex,
+        int inputProviderParamIndex,
+        int outputReceiverParamIndex) {
+      this.targetMethod = targetMethod;
+      this.boundedWindowParamIndex = boundedWindowParamIndex;
+      this.inputProviderParamIndex = inputProviderParamIndex;
+      this.outputReceiverParamIndex = outputReceiverParamIndex;
+    }
+
+    /**
+     * Utility method: returns how many parameters the method has, not counting the {@link
+     * DoFn.ProcessContext} parameter.
+     */
+    public int numExtraParameters() {
+      int res = 0;
+      res += (boundedWindowParamIndex == -1) ? 0 : 1;
+      res += (inputProviderParamIndex == -1) ? 0 : 1;
+      res += (outputReceiverParamIndex == -1) ? 0 : 1;
+      return res;
+    }
+  }
+
+  /** Describes a {@link DoFn.StartBundle} or {@link DoFn.FinishBundle} method. */
+  public static class BundleMethod {
```

Comment:
I would propose a single:

`public static class DoFnMethod { ... }`

which encapsulates the common logic of having a `Method`, and a list of extra arguments.

That will also make it easier for the future additions (such as adding `@OnTimer` methods), etc.